### PR TITLE
CMake build: Allow building with C++11/14 (use e.g. -D abi=c++11)

### PR DESCRIPTION
**Pull request by jpetso**
I was originally reluctant to submit this as pull request, but then I saw that in the b2 build you've also got support and special handling for non-default standards. Since some Boost headers change their ABI based on C++ standard and libtorrent pulls in a number of those through public headers, this added CMake option has been coming in handy to allow building with other C++11 code.

### CODE REVIEW

**Comment by jpetso** [comment]
Trivia: This is also the last commit on my fork before I can switch back to upstream master - all the other stuff has been fixed or improved so I was able to unfork my branch. Well done!

File: CMakeLists.txt [hunk 1]
```diff
@@ -306,6 +314,15 @@ else()
 	endif (MSVC)
 endif()
 
+if (NOT MSVC)
+	if (abi STREQUAL "c++03")
+		add_definitions(-std=${abi})
+	elseif (abi STREQUAL "c++11" OR abi STREQUAL "c++14")
+		# Also silence auto_ptr because it throws lots of warnings.
+		add_definitions(-std=${abi} -Wno-deprecated-declarations)
```
Comment (arvidn) [inline CMakeLists.txt]: what warnings do you get about auto pointers?

File: CMakeLists.txt [hunk 2]
```diff
@@ -317,7 +334,9 @@ if (MSVC)
 #$(SolutionDir)<toolset>msvc,<variant>release:<linkflags>/OPT:ICF=5
 #$(SolutionDir)<toolset>msvc,<variant>release:<linkflags>/OPT:REF
 elseif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
-	add_definitions(-Wno-c++11-extensions)
+	if (abi STREQUAL "c++03")
```
Comment (arvidn) [inline CMakeLists.txt]: what about c++98?

File: CMakeLists.txt [hunk 3]
```diff
@@ -161,10 +161,18 @@ option(dht "enable support for Mainline DHT" ON)
 option(resolve-countries "enable support for resolving countries from peer IPs" ON)
 option(unicode "enable unicode support" ON)
 option(deprecated-functions "enable deprecated functions for backwards compatibility" ON)
+if (NOT MSVC)
+	# MSVC doesn't provide this option and has its own ABI issues anyway.
+	option(abi "pick one until libtorrent's ABI is robust: c++03, c++11, c++14, default (currently c++03)" "default")
```
Comment (arvidn) [inline CMakeLists.txt]: how is this related to the robustness of libtorrent's ABI?

File: CMakeLists.txt [hunk 4]
```diff
@@ -306,6 +314,15 @@ else()
 	endif (MSVC)
 endif()
 
+if (NOT MSVC)
+	if (abi STREQUAL "c++03")
+		add_definitions(-std=${abi})
```
Comment (arvidn) [inline CMakeLists.txt]: I guess the only compilers are GCC and MSVC? :/

Comment (arvidn) [inline CMakeLists.txt]: the CMakeLists.txt probably hasn't worked with anything other than gcc compatible and msvc compilers before either though.

Comment (jpetso) [inline CMakeLists.txt]: In practice, you likely get better compatibility assuming that a compiler is GCC compatible than only supporting known ones. ARM toolchain and ICC both use GCC basic syntax for these kinds of things, and it's easier to add exceptions for the cases where a certain compiler doesn't support a given option.

In general, I'm a fan of writing things in a generic way, but without an easy way to test it, I think it's reasonable to stick to GCC, Clang and MSVC as requirements for a build script change.

File: CMakeLists.txt (hunk 3 above)
Comment (jpetso) [inline CMakeLists.txt]: That's actually a very good question, and I might have to check if this is actually still relevant now that asio is built header-only. Let me get back to you on this.

File: CMakeLists.txt (hunk 2 above)
Comment (jpetso) [inline CMakeLists.txt]: According to Wikipedia, C++03 is just a more polished wording of C++98, so features provided by both should be the same and Boost wouldn't change their definitions based on that. So I picked the newer one since it's better defined and certainly (hopefully?) old enough for any libtorrent users to have supported by their compiler.

File: CMakeLists.txt (hunk 1 above)
Comment (jpetso) [inline CMakeLists.txt]: `auto_ptr` is deprecated in C++11, and libtorrent::session uses it for dispatch handlers. That generates a ton of deprecation warnings.

Comment (arvidn) [inline CMakeLists.txt]: those warnings are supposed to be suppressed. at least for msvc, gcc and clang. is that broken? or do you use a different compiler?

Comment (arvidn) [inline CMakeLists.txt]: (I would really prefer to not disable _all_ deprecated declarations)

File: CMakeLists.txt (hunk 2 above)
Comment (arvidn) [inline CMakeLists.txt]: my understanding is that the only difference is that more defect-reports have been resolved and TR1 was included in the std::tr1 namespace. I think it's still commonly referred to as c++98 though, since that's still the version of the ISO standards document, maybe both strings could be added as synonyms?

Perhaps I'm not quite understanding the motivation here though. The issue is that boost changes its ABI depending on whether it's built with a c++11 compiler or not, right? So, in order to stay ABI compatible, libtorrent must be built with the same version of the standard, so it must be configurable? (would you mind mentioning this in building.rst, or somewhere appropriate too?)

File: CMakeLists.txt (hunk 3 above)
Comment (arvidn) [inline CMakeLists.txt]: on a side-note, I ran into a nasty issue now that asio is built header's only. https://github.com/arvidn/libtorrent/blob/master/simulation/test_http_connection.cpp#L130

File: CMakeLists.txt (hunk 4 above)
Comment (arvidn) [inline CMakeLists.txt]: I didn't realize gcc supported -std=c++03

**Comment by LordNyriox** [comment]
@arvidn Any chance this might get merged?  It's an old PR and might not be mergeable in the future. 

Thanks, 
~ Nyriox

**Comment by arvidn** [comment]
I think most of this patch is fine. there are still a few outstanding questions though (see the comments)

**Comment by arvidn** [comment]
also, it might make sense to move this PR over to be against RC_1_1 now

**Comment by LordNyriox** [comment]
@jpetso Do you think you might be able to update this PR?  @arvidn seems to want this for the next release.  

Thanks, 
~ Nyriox

**Comment by jpetso** [comment]
@LordNyriox I'll have to try building and testing my codebase (C++11, using boost with C++11 ABI) with (pre-C++11) default libtorrent builds to see if this patch is still needed. I can't commit to any particular deadline (sorry) but I'll see if I can check it before the weekend. Call me out on it if I still haven't managed to take the time by then.

**Comment by LordNyriox** [comment]
@jpetso All right.  Thanks for replying! 

~ Nyriox

**Comment by LordNyriox** [comment]
@jpetso Just a reminder, it is currently Saturday, February 13, and I have not yet seen a report on your progress.  

Thanks for letting me remind you. :} 

~ Nyriox

**Comment by jpetso** [comment]
Well... at least I got a lot of other stuff done on the weekend, unrelated to libtorrent or even software in general. :blush: 

Sorry.

**Comment by LordNyriox** [comment]
@jpetso: Don't worry about it.  Other than needing to switch target branches to RC_1_1, I don't think this PR could do any harm.  

@arvidn:  I'm not sure much maintenance is going to be done on this PR, so I think any backporting will have to be done by you.  : {   Other than that, I think this is as ready to be merged as it'll ever be.  

Thanks, 
~ Nyriox

**Comment by MikaelSmith** [comment]
This may be unnecessary, as you can specify `-DCMAKE_CXX_STANDARD=14` when invoking CMake.

**Comment by LordNyriox** [comment]
@arvidn:  I think you should go ahead and merge this.  You can revert any concerning code in a later commit.

**Comment by arvidn** [comment]
with master now requiring c++11, and gcc6 presumably not being in widespread use yet (where the default is c++14 instead of 98). Is specifying CMAKE_CXX_STANDARD the typical cmake-way of dealing with this? assuming that the compiler takes a "-std=c++11" flag seems a bit presumptuous, but I don't get the impression cmake abstracts compilers in that way.

**Comment by jpetso** [comment]
CMAKE_CXX_STANDARD seems to be the standard way to do it, however it's not available until CMake 3.1.

**Comment by jpetso** [comment]
Closing since there's a good chance that PR #682 makes this obsolete.
